src/main/java/uk/nhs/hee/trainee/details/config/SchedulePublishRetryConfig.java

```diff
@@ -0,0 +1,124 @@
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.hee.trainee.details.config;
+
+import java.util.Collection;
+import java.util.Queue;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.amqp.rabbit.connection.CorrelationData;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+import uk.nhs.hee.trainee.details.model.ProgrammeMembership;
+import uk.nhs.hee.trainee.details.model.RetryCorrelationData;
+import uk.nhs.hee.trainee.details.service.RabbitPublishService;
+
+/**
+ * Configure Rabbit message republishing for not-acknowledged and unconfirmed errors.
+ */
+@Slf4j
+@Configuration
+@EnableScheduling
+public class SchedulePublishRetryConfig {
+
+  private final RabbitTemplate rabbitTemplate;
+  private final RabbitPublishService rabbitPublishService;
+  private final int maxRetries;
+
+  /**
+   * Initialise the SchedulePublishRetryConfig object.
+   *
+   * @param rabbitTemplate       the Rabbit template.
+   * @param rabbitPublishService the Rabbit publishing service.
+   * @param maxRetries           the maximum number of message retries.
+   */
+  public SchedulePublishRetryConfig(RabbitTemplate rabbitTemplate,
+                                    RabbitPublishService rabbitPublishService,
+                                    @Value("${spring.rabbitmq.max-retries}") int maxRetries) {
+    this.rabbitTemplate = rabbitTemplate;
+    this.rabbitPublishService = rabbitPublishService;
+    this.maxRetries = maxRetries;
+  }
+
+  /**
+   * Republish not-acknowledged messages (up to a maximum number of retries).
+   */
+  @Scheduled(fixedDelay = 10000)
+  public void scheduleNackedRepublishTask() {
+    Queue<RetryCorrelationData> queue = rabbitPublishService.getNegativeAckedMessages();
+
+    // try re-publish with batch size of 100 each time
+    for (int i = 0; i < 100; i++) {
+      if (queue.isEmpty()) {
+        break;
+      }
+      RetryCorrelationData retryCorrelationData = queue.remove();
+
+      log.info("Retry nack-ed Rabbit message id : '{}'", retryCorrelationData.getId());
+      checkRetryCountAndRepublish(retryCorrelationData.getRetryCount(),
+          retryCorrelationData.getId());
+    }
+  }
+
+  /**
+   * Republish unconfirmed messages (up to a maximum number of retries).
+   */
+  @Scheduled(fixedDelay = 5000)
+  public void scheduleUnconfirmedRepublishTask() {
```

Comment:
Also non-blocking. I haven't yet actually triggered this scenario - some investigation would be needed as to how to do this.

Reply:
Can be done by starving rabbit of resources by e.g. setting `rabbitmqctl set_vm_memory_high_watermark 0.005` in the container